source/unified-test-format/unified-test-format.rst

```diff
@@ -1170,6 +1176,14 @@ that the field is set and is a non-empty BSON ObjectId (i.e. all bytes of the
 ObjectId are not 0). If false, test runners MUST assert that the field is not
 set or is an empty BSON ObjectId.
 
+hasServerConnectionId
+`````````````````````
+
+This field is an optional boolean that specifies whether or not the
+``serverConnectionId`` field of an event is set. If true, test runners MUST
+assert that the field is set and is an Int64. If false, test runners MUST assert
```

Comment:
Is there any restriction on the value if the field is set (i.e. non-zero)? Just curious if there's relevance for drivers that might not actually implement the field as an optional type. For example, libmongoc might want to return an `int64_t` and use zero (or a constant defined to -1) to signify an unset value, rather than have to create a struct that represents an optional int64_t like it did for booleans in Versioned API.

Perhaps this comment is more relevant to the APM spec itself where the fields are documented.

Reply:
The [underlying type](https://github.com/mongodb/mongo/blob/master/src/mongo/db/client.h#L60) of `ConnectionId` in the server code is a `long long int`, so `serverConnectionId` _could_ be <= 0. That said, judging from [this internal logic](https://github.com/mongodb/mongo/blob/master/src/mongo/db/client.h#L184-L189), I believe a connection to an external user (i.e. driver) invariantly has a positive `serverConnectionId`. A `ConnectionId` of `0` seems to be used for internal system connections.

I've added some wiggle room for drivers without native Optional-like structures similar to the qualifications of  `hasServiceId` above.

Reply:
Looking at the `hello` response, the server is returning the value as a BSON Int32.  In practice, it's hard to imagine ever needing a value greater than max Int32, but I suppose we can be flexible, but FYI Java driver (which implemented this a long time ago) models the connectionId as an Int32.

Reply:
It looks like the server is returning a BSON Int32 despite it being stored as a `long long int` server-side. I also don't think the value will exceed max Int32, so I've switched to an Int32 instead of Int64. That way, the Java driver is already compliant with the spec, too. If anyone else has a strong opinion against I'll leave this thread open.

Reply:
Int32 sounds fine to me. If we codify that in a spec, it gives us an excuse should the server ever decide to deviate from its current behavior (server folks do reference our specs from time to time). You can easily include a note that drivers MAY use a wider type to represent the value but the server's behavior is to return an Int32.

Reply:
That sounds good. I've included a note that drivers may use a wider type.